File: Easy/412/412.go
func fizzBuzz(n int) []string {
    answer := make([]string, 0, n)

    for i := 1; i <= n; i++ {
        if i%3 == 0 && i%5 == 0 {
            answer = append(answer, "FizzBuzz")
        } else if i%3 == 0 {
            answer = append(answer, "Fizz")
        } else if i%5 == 0 {
            answer = append(answer, "Buzz")
        } else {
            answer = append(answer, strconv.Itoa(i))
        }
    }
    return answer
}

File: 412/412.java
class Solution {
    public List<String> fizzBuzz(int n) {
        List<String> answer = new ArrayList<>();

        for (int i = 1; i <= n; i++) {
            if (i % 3 == 0 && i % 5 == 0 ) {
                answer.add("FizzBuzz");
            } else if ( i % 3 == 0 ) {
                answer.add("Fizz");
            } else if ( i % 5 == 0 ) {
                answer.add("Buzz");
            } else {
                answer.add(String.valueOf(i));
            }
        }
        return answer;
    }
}


// ---------------->>> * <<< -------------------------\\
// revised solution

class Solution2 {
    public List<String> fizzBuzz(int n) {
        List<String> answer = new ArrayList<>();

        for (int i = 1; i <= n; i++) {
            boolean divisibleBy3 = i % 3 == 0;
            boolean divisibleBy5 = i % 5 == 0;

            String currStr = "";

            if (divisibleBy3) {
                currStr += "Fizz";
            }
            if (divisibleBy5) {
                currStr += "Buzz";
            }
            if (currStr.isEmpty()) {
                currStr += String.valueOf(i);
            }
            answer.add(currStr);
        }
        return answer;
    } 
    // Time complexity = O(n)
    //Space complexity = O(1)
}
